File: app/src/main/java/com/raju/javabaseproject/dagger/modules/base/NetworkModule.java
package com.raju.javabaseproject.dagger.modules.base;

import android.content.Context;
import android.support.annotation.Nullable;

import com.raju.javabaseproject.R;
import com.raju.javabaseproject.data.http.HeaderInterceptor;
import com.raju.javabaseproject.data.http.HttpClientBuilder;
import com.raju.javabaseproject.data.utils.SharedPrefsHelper;
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Rajashekhar Vanahalli on 05/04/18.
 */

@Module(includes = HttpModule.class)
public class NetworkModule {

    public NetworkModule() {

    }

    @Provides
    @Singleton
    OkHttpClient providesOkHttpClient(Context context, Interceptor networkInterceptor, @Named("LOGGING") Interceptor loggingInterceptor,
                                      SharedPrefsHelper sharedPrefsHelper) {
        return buildOkHttpClient(context, context.getString(R.string.app_name), 50, networkInterceptor, loggingInterceptor,
                new HeaderInterceptor(sharedPrefsHelper));
    }

    @Provides
    @Singleton
    Retrofit providesRetrofit(Context context, OkHttpClient okHttpClient) {
        Retrofit retrofit = new Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(context.getResources().getString(R.string.endpoint))
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        return retrofit;
    }

    private OkHttpClient buildOkHttpClient(Context context, String cacheName, int cacheSize, @Nullable Interceptor networkInterceptor,
                                           @Nullable Interceptor loggingInterceptor, @Nullable Interceptor headerInterceptor) {
        HttpClientBuilder builder =
                new HttpClientBuilder(context)
                        .setTimeouts(120)
                        .setCache(cacheName, cacheSize)
                        .setLoggingInterceptor(loggingInterceptor)
                        .setWireFormatHeaderInterceptor(headerInterceptor)
                        .enableSocketLog(true);

        return builder.build();
    }
}

File: app/src/main/java/com/raju/javabaseproject/dagger/modules/base/NetworkModule.kt
package com.raju.javabaseproject.dagger.modules.base

import android.content.Context

import com.raju.javabaseproject.R
import com.raju.javabaseproject.data.source.remote.http.HeaderInterceptor
import com.raju.javabaseproject.data.source.remote.http.HttpClientBuilder
import com.raju.javabaseproject.data.utils.SharedPrefsHelper
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory

import javax.inject.Named
import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

@Module(includes = [HttpModule::class])
class NetworkModule {

    @Provides
    @Singleton
    internal fun providesOkHttpClient(context: Context, networkInterceptor: Interceptor, @Named("LOGGING") loggingInterceptor: Interceptor,
                                      sharedPrefsHelper: SharedPrefsHelper): OkHttpClient {
        return buildOkHttpClient(context, context.getString(R.string.app_name), 50, networkInterceptor, loggingInterceptor,
                HeaderInterceptor(sharedPrefsHelper))
    }

    @Provides
    @Singleton
    internal fun providesRetrofit(context: Context, okHttpClient: OkHttpClient): Retrofit {
        val retrofit = Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(context.resources.getString(R.string.endpoint))
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()

        return retrofit
    }

    private fun buildOkHttpClient(context: Context, cacheName: String, cacheSize: Int, networkInterceptor: Interceptor?,
                                  loggingInterceptor: Interceptor?, headerInterceptor: Interceptor?): OkHttpClient {
        val builder = HttpClientBuilder(context)
                .setTimeouts(120)
                .setCache(cacheName, cacheSize)
                .setLoggingInterceptor(loggingInterceptor!!)
                .setWireFormatHeaderInterceptor(headerInterceptor!!)
                .enableSocketLog(true)

        return builder.build()
    }
}
